designs/2022-languages/README.md

```diff
@@ -0,0 +1,612 @@
+        // ...
+    }
+};
+```
+
+Here, `lang1` and `lang2` are languages defined by the plugin. End users can specify which language to use via the `language` key in a config object, such as:
+
+
+```js
+// inside eslint.config.js
+import lang from "eslint-plugin-lang";
+
+exports default [
+    {
+        plugins: {
+            lang
+        },
+        files: ["**/*.lang"],
+        language: "lang/lang1",
+        languageOptions: {
+            // options
+        }
+    }
+];
+```
+
+Here, `"lang/lang1"` refers to the `lang1` language as exported from `eslint-plugin-lang`. Anything inside of `languageOptions` is then handled by `lang1` rather than by ESLint itself. 
+
+### Language Definition Object
+
+Each language definition object must implement the following interface:
+
+```ts
+interface ESLintLanguage {
+
+    /**
+     * Indicates how ESLint should read the file.
+     */
+    fileType: "text" | "binary";
+
+    /**
+     * Validates languageOptions for this language.
+     */
+    validateOptions(options: LanguageOptions): void;
+
+    /**
+     * Parses the given file input into its component parts.
+     */
+    parse(file: File, env: LanguageContext): ParseResult | Promise<ParseResult>;
+
+    /**
+     * Creates SourceCode object that ESLint uses to work with a file.
+     */
+    createSourceCode(file: File, input: ParseResult, env: LanguageContext): SourceCode | Promise<SourceCode>;
+
+}
+
+// Supplemental interfaces
+
+interface LanguageOptions {
+    // language-defined
+}
+
+interface File {
+
+    /**
+     * The path that ESLint uses for this file. May be a virtual path
+     * if it was returned by a processor.
+     */
+    path: string;
+
+    /**
+     * The path to the file on disk. This always maps directly to a file
+     * regardless of whether it was returned from a processor.
+     */
+    physicalPath: string;
+
+    /**
+     * Indicates if the original source contained a byte-order marker.
+     * ESLint strips the BOM from the `body`, but this info is needed
+     * to correctly apply autofixing.
+     */
+    bom: boolean;
+
+    /**
+     * The body of the file to parse.
+     */
+    body: string | ArrayBuffer
+}
+
+interface LanguageContext {
+    options: LanguageOptions;
+}
+
+interface ParseResult {
+    ast: ASTNode;
+    body: string | ArrayBuffer;
+}
+```
+
+At a high-level, ESLint uses the methods on a language object in the following way:
+
+* During config validation, ESLint calls `validateOptions()` on the `languageOptions` specified in the config. It is the expectation that 
+* When ESLint reads a file, it checks `fileType` to determine whether the language would prefer the file in text or binary format.
+* After reading a file, ESLint passes the file information to `parse()` to create a raw parse result.
+* The raw parse result is then passed into `createSourceCode()` to create the `SourceCode` object that ESLint will use to interact with the code.
+
+#### The `validateOptions()` Method
```

Comment:
Perhaps we could also allow customizing merging behavior. In fact, instead of the `validateOptions()` method, this could be a `languageOptionsSchema` property. We could then use the provided schema in place of the current [`languageOptions`](https://github.com/eslint/eslint/blob/6380c87c563be5dc78ce0ddd5c7409aaf71692bb/lib/config/flat-config-schema.js#L445-L453) schema.

Reply:
For convenience, we could move [`deepMerge`](https://github.com/eslint/eslint/blob/3e34418b31664decfb2337de798feafbf985b66c/lib/config/flat-config-schema.js#L60) (I assume this merge strategy was intended for language options) into [`ObjectSchema`](https://github.com/humanwhocodes/object-schema/blob/main/src/merge-strategy.js). Then, specifying a schema that just validates the object as a whole wouldn't be much more complicated than specifying a `validateOptions()` method:

```js
// An ESLintLanguage implementation
{
    // ...

    languageOptionsSchema: {
        merge: "deepAssign",
        validate(options) {
            // ....
        }        
    }
}
```